### PR TITLE
Add a map object

### PR DESCRIPTION
**Pull request by j-f1**
The map will automatically fill in as you explore Ainneve. Each time you enter a room, the map will look through all of the exits representing a compass direction and store information about the rooms on the other side. The `map` command (or `look map`) will show you your map. By default, the map is also printed every time you move to a new room. The webclient has the option to show the map in a popup window, kind of like the `help` command. In addition, I’ve:

- removed an impossible exit from the meadow which really messed up the map
- added a helper function to allow specifying terrains based on the hardness levels
- added grass and road terrains
- added an `@map_tile` command to customize the tile shown on the map for a specific room
- added `Road` and `Bridge` typeclasses that render connected lines on the map. Example:
  ```
    ┌─══
  ╶─┘
  ```
  consists of:
  ```
        (road)(bridge)
  (road)(road)
  ```
- gives each character a map when they finish calling `@charcreate`

The `map` command also includes a `map/reset` command, which erases all of the data on the map, which is useful when testing the map.

Here’s the current world map:

```
┌───── MAP ────┐
│  . ~↟~ .     │
│     . .↟. .  │
│     .  .  .  │
│  ══─*─────╴  │
│     \        │
│       \^_    │
└──────────────┘
```

### CODE REVIEW

File: commands/maps.py
Comment (aliceafterall) [inline commands/maps.py]: Perhaps document what each argument of get_input is doing? Ainneve is meant to be an example, so it might help those new to Evennia.

File: commands/maps.py
Comment (aliceafterall) [inline commands/maps.py]: Same type of thing: It may be best to split this to two lines. I know what it does, but someone brand new to Python and Evennia might not.

File: commands/room_exit.py
Comment (aliceafterall) [inline commands/room_exit.py]: You shouldn't take out the acceptable terrain message - that's actually for the command help, we shouldn't have to read the source to know. :]

File: typeclasses/maps.py
Comment (aliceafterall) [inline typeclasses/maps.py]: Same as above - may wish to split to two lines.

File: typeclasses/maps.py
Comment (aliceafterall) [inline typeclasses/maps.py]: I really don't like this loop - if I make a section of rooms containing themselves:
`Room1 -> Room2 -> Room3 -> Room1`
Then this will loop forever. Yes, that shouldn't happen but still.

File: typeclasses/maps.py
Comment (aliceafterall) [inline typeclasses/maps.py]: Same thing here, I might suggest a way to break out of the loop after some amount of iterations. I don't think this is as likely to cause problems though.

File: typeclasses/maps.py
Comment (aliceafterall) [inline typeclasses/maps.py]: Lines 93 and 94 should possibly be split to two lines each.
Each of the if statements should also possibly be split:
```
if x < min_x:
    min_x = x
```
Again, for clarity and easy reading.

File: typeclasses/maps.py
Comment (aliceafterall) [inline typeclasses/maps.py]: This seems like it could be simplified? It's hard to understand this way.

File: typeclasses/rooms.py
Comment (aliceafterall) [inline typeclasses/rooms.py:155]: Have you tested this on various clients? I wonder how it looks on older clients.

**Review by aliceafterall — CHANGES_REQUESTED** [review]
Just some readability and simplification things - this is awesome stuff you've written!

File: commands/room_exit.py
Comment (j-f1) [reply commands/room_exit.py]: Do you know if there’s a way to dynamically load the keys? Otherwise, the list could get out-of-date.

File: typeclasses/maps.py
Comment (j-f1) [reply typeclasses/maps.py]: Hm. I could do `title.center(width, '#')` and replace all of the `#` characters with the box drawing character.

File: typeclasses/rooms.py
Comment (j-f1) [reply typeclasses/rooms.py:155]: I haven’t. This should work anywhere Unicode is supported, though. If you’d like, I could potentially add an option to use boring ASCII characters where possible.

File: commands/room_exit.py
Comment (aliceafterall) [reply commands/room_exit.py]: Good point! Let me look into it and see.

File: typeclasses/maps.py
Comment (aliceafterall) [reply typeclasses/maps.py]: Hmm. That might work. Isn't there an evennia util as well?

File: typeclasses/rooms.py
Comment (aliceafterall) [reply typeclasses/rooms.py:155]: You know, rather than speculate I'll try it out and get back to you! :)

File: typeclasses/maps.py
Comment (j-f1) [reply typeclasses/maps.py]: It looks like the Evennia `utils.utils.pad` just calls `str.center`.

File: commands/room_exit.py
Comment (j-f1) [reply commands/room_exit.py]: @chainsol Any updates here?

Comment (aliceafterall) [reply commands/room_exit.py]: Sorry, was away from my computer!
What if we used _TERRAIN.keys()?
", ".join( _TERRAIN.keys() )

**Comment by j-f1** [comment]
Should I contribute this instead to the Evennia contrib? It seems like it isn’t specific to Ainneve.

**Comment by aliceafterall** [comment]
I agree - it's really quite useful for most any game. If you try it on Ainneve and it works well, perhaps a contrib for Evennia?

File: typeclasses/rooms.py
Comment (aliceafterall) [reply typeclasses/rooms.py:155]: If there's a way to check if Unicode characters are able to be used, it would be good to have an alternative. If the player doesn't support Unicode characters, won't these either be stripped or display as boxes? It seems like Windows telnet doesn't support Unicode, same for some Android clients. Perhaps it doesn't matter?

Comment (j-f1) [reply typeclasses/rooms.py:155]: Do you know if there’s a way to tell if a session supports Unicode?

Comment (aliceafterall) [reply typeclasses/rooms.py:155]: Hmm. You can check the encoding setting the player sets - maybe if it's anything other than utf-8, use a fallback? I'd also suggest in the future some form of screenreader mode if you didn't already consider that.

Comment (j-f1) [reply typeclasses/rooms.py:155]: I’m not exactly sure what the fallback should consist of. Here’s what I’ve thought of so far:

```python
FALLBACK_CENTERS = {
        'n': '|',
        'e': '-',
        's': '|',
        'w': '-',
      'e n': '?',
      'n s': '|',
      'n w': '?',
      'e s': '?',
      'e w': '-',
      's w': '?',
    'e n w': '?',
    'e s w': '?',
    'n s w': '?',
    'e n s': '?',
  'e n s w': '+'
}
```

Comment (aliceafterall) [reply typeclasses/rooms.py:155]: Hmm. Would slashes ( / and \ ) work? I'm not sure about the directional crossroads though.

**Comment by j-f1** [comment]
Can I get some 👀 on this @chainsol?

**Review by aliceafterall — APPROVED** [review]

**Comment by aliceafterall** [comment]
Let me try it on my little test server, just to make sure it runs on 0.7.

**Comment by aliceafterall** [comment]
Actually, I already have some concerns - I don't like the idea of modifying the webclient like this, and it already breaks the webclient.

Edit: To be clear to anyone reading this later - it's because Evennia 0.7 loads jQuery from the CDN, not the static folder.

File: web/template_overrides/webclient/webclient.html
Comment (aliceafterall) [inline web/template_overrides/webclient/webclient.html]: jQuery isn't loaded directly from the js folder - it's loaded via the code.jquery.com CDN, so this breaks the webclient

**Comment by j-f1** [comment]
How does this break the webclient?

File: web/template_overrides/webclient/webclient.html
Comment (aliceafterall) [inline web/template_overrides/webclient/webclient.html:65]: In general I don't like the idea of adding overrides to the webclient like this - I'd want to see what @Griatch and @feend78 think of this.

**Review by aliceafterall — CHANGES_REQUESTED** [review]
Overall everything looks good - I'm wary of the webclient having a map interface, because updates to Evennia core that improve the webclient will cause problems in the future, or we'll not get them.

I still think having a fallback for the map display is important, and you had some good thoughts on that.

**Comment by aliceafterall** [comment]
It's because Evennia core, at least in 0.7, doesn't store jQuery, it's loaded from the CDN. Changing that line of course fixes it.

File: web/template_overrides/webclient/webclient.html
Comment (j-f1) [reply web/template_overrides/webclient/webclient.html]: Fixed!

**Comment by j-f1** [comment]
I’m wondering if it’s possible to install the map logic by overwriting the `onText` function in a separate script. I’ll try to move the other (notification-related) change to a PR on evennia/evennia this weekend.

**Comment by aliceafterall** [comment]
Nice. I'm going to wait for @Griatch or @feend78 to chime in before I approve just because I want some feedback from others about the webclient thing. I'm just wary of it...

Edit: Thinking about this, it would actually be good to be able to add "plugins" like this to the webclient anyway. Perhaps the core Evennia could support something like that...

File: typeclasses/maps.py
Comment (aliceafterall) [inline typeclasses/maps.py]: Ah, missed this. It should be current_character.account - since the name changed.

**Comment by aliceafterall** [comment]
@j-f1 I just noticed - line 70 in typeclasses/maps.py, you check the character.player, but we'll have renamed that to character.account.

File: typeclasses/rooms.py
Comment (Griatch) [reply typeclasses/rooms.py:155]: FYI, the `UTF-8` flag is set on the client session if the client reports that it supports UTF-8. It is visible via `@options` and in code via e.g. `character.sessions.get()[0].session_flags`.

**Comment by aliceafterall** [comment]
@j-f1 I'd like to merge this into Develop, is it possible to fix any references to the player to match Develop, as well as to update the dictionary we talked about, of alternate characters for non-unicode clients?

**Comment by j-f1** [comment]
I’m hoping to get to it this weekend, although I might have a chance sooner than that.

**Comment by aliceafterall** [comment]
Of course, no rush! I just wanted you to know I hadn't forgotten!

File: typeclasses/maps.py
Comment (j-f1) [inline typeclasses/maps.py:60]: Should I change this to something other than `show_player`? When set to `True`, it puts a `*` on the map in the room you’re in.

Comment (aliceafterall) [reply typeclasses/maps.py:60]: To me, `show_player` makes sense - I don't see a problem with that naming.

File: web/template_overrides/webclient/webclient.html
Comment (j-f1) [reply web/template_overrides/webclient/webclient.html:65]: @Griatch or @feend78 Could you comment on this?

**Comment by aliceafterall** [comment]
@j-f1 Would you say you're ready with this? The change to a separate file for the webclient changes eases my qualms with this, and I can't see anything wrong!

**Review by aliceafterall — APPROVED** [review]

**Comment by j-f1** [comment]
@chainsol I’m good to go if you’re ready. Should I worry about [support for non-Unicode clients](https://github.com/evennia/ainneve/pull/97#discussion_r129674641)?

**Comment by aliceafterall** [comment]
Yes, if you're able to, I think that would be important - sounds like Griatch suggested `character.sessions.get()[0].protocol_flags['ENCODING']` as a way to check for UTF-8/Unicode support.

**Comment by aliceafterall** [comment]
@j-f1 Would it be doable to just change the map tiles to a tuple, with the unicode character and a fallback character, and then check if the session supports Unicode in the Map object itself?

**Comment by j-f1** [comment]
That’s a good idea! I’ll take a look at that this afternoon. (~5-6 hours from now)

**Comment by aliceafterall** [comment]
No rush! When you get to it, great, and we'll merge this in to the develop branch.

**Comment by j-f1** [comment]
I’ve looked at this a few times, but I’m not sure how I should thread the result down to the `map_tile` methods.

**Comment by aliceafterall** [comment]
Hmm, I've been going crazy moving, but I see your troubles there. Could the map_tile functions be passed the player trying to look at them?

Edit: Oh goodness, did you mean the `@map_tile` command? Or the map_tile method on the rooms?

**Comment by j-f1** [comment]
> Oh goodness, did you mean the `@map_tile` command? Or the map_tile method on the rooms?

I meant the `map_tile` method on rooms.

**Comment by aliceafterall** [comment]
So would that work? Pass down from the commands the session of the player looking at the map, and default to showing UTF-8/Unicode characters, fall-back to ASCII if the session doesn't support UTF-8?

**Comment by aliceafterall** [comment]
```"
ââââ MAP ââââââ
â  . ~â~ .     â
â     . .â. .  â
â     *  .  .  â
â  ââââ¬ââââââ´  â
â     \        â
â       \^_    â
ââââââââââââââââ
```
Without having fallback characters, that's what the map looks like. I think we need to figure this out before merging it in. With a `utf-8` encoding, it looks great, though!

I took a look at it and if we changed the character dicts to be storing tuples of the "nice" character and a fallback, that might be best? That way we can check if a character is holding the map and if so, use the fallback character if the session doesn't support `utf-8`.

**Pull request closed without merging**
